include/opendht/peer_discovery.h

```diff
@@ -0,0 +1,208 @@
+/*
+ *  Copyright (C) 2014-2019 Savoir-faire Linux Inc.
+ *  Author(s) : Mingrui Zhang <mingrui.zhang@savoirfairelinux.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "sockaddr.h"
+#include "infohash.h"
+
+#ifdef _WIN32
+#include <Winsock2.h> // before Windows.h, else Winsock 1 conflict
+#include <Ws2tcpip.h> // needed for ip_mreq definition for multicast
+#include <Windows.h>
+#else
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#endif
+
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h> 
+
+#include <thread>
+#include <string>
+#include <mutex>
+#include <condition_variable>
+
+namespace dht {
+
+class OPENDHT_PUBLIC PeerDiscovery
+{
+public:
+
+    using PeerDiscoveredCallback = std::function<void(const InfoHash&, const SockAddr&)>;
+
+    PeerDiscovery(sa_family_t domain, in_port_t port);
+    ~PeerDiscovery();
+    
+    /**
+     * startDiscovery - Listen
+    */
+    void startDiscovery(PeerDiscoveredCallback callback, in_port_t port_to_avoid);
+
+    /**
+     * startPublish - Send
+    */
+    void startPublish(dht::InfoHash nodeId, in_port_t port_to_send);
```

Comment:
These parameters will be passed by copy, I think you can replace a lot of arguments in prototypes by `T& foo` or even `const T& foo` to avoid any copy.